parse.go

```diff
@@ -8,9 +8,25 @@ import (
 	"strings"
 )
 
+const (
+	// DefaultBlockSize is the chunk size that splitters produce (or aim to).
+	DefaultBlockSize int64 = 1024 * 256
+
+	// 1 MB, on-wire block size for "datablocks ( unixfs, etc )"
+	// copy of https://github.com/ipfs/go-unixfs/blob/v0.2.3/importer/helpers/helpers.go#L8
+	BlockSizeLimit int = 1048576
+
+	// in case we are using raw-leaves: this would match BlockSizeLimit, but we can't assume that
+	// be conservative and substract the PB wraping size of a full DAG-PB+UnixFS node describing 1M
+	// (2b(type2/file)+4b(data-field:3-byte-len-delimited)+4b(size-field:3-byte-varint))+(4b(DAG-type-1:3-byte-len-delimited))
+	// FIXME - this calculation will need an update for CBOR
+	BlockPayloadLimit int = (BlockSizeLimit - (2 + 4 + 4 + 4))
+)
+
```

Comment:
@Stebalien @dirkmc specifically want your 👍 that  the thinking behind this const-block is legit

Reply:
Note: the real "sane" upper limit for IPLD blocks is probably more like 2MiB (the network limit is actually 4MiB). As long as we never create chunks over 1MiB, we should be safe.

@mikeal have we made any progress specing maximum block sizes?

Reply:
@Stebalien even if we went with 2MiB for IPLD the spot here is constrained by https://github.com/ipfs/go-unixfs/blob/v0.2.3/importer/helpers/helpers.go#L8, so... for the context of this PR / issue: are these values correct, or we are looking at updating the importer too...?

Reply:
In practice, that's the maximum _chunk_ size. It was probably intended to be the maximum block size but I'm concerned that lowering the maximum chunk size will break end users.

I'd keep that as the maximum chunk size for now and set the max _block_ size to 2MiB (or even 1.25MiB?).

Reply:
I am confused...
In this PR I really only use `BlockPayloadLimit` ( what you called "maximum chunk size" )
We seem to agree that 1MiB - wrapping is reasonable (?)

The constant of `BlockSizeLimit` is there **solely** to derive BlockPayloadLimit, I am not using it in other ways.
Perhaps I should just remove it and write out 1MiB with a comment "matching xxx"

Would that work?

Reply:
>  We seem to agree that 1MiB - $wrapping is reasonable (?)

I think that's actually our disagreement. I want users to be able to remember the maximum chunk size, and I don't want it to depend on things like how big the padding is (could change, e.g., with metadata). I'd like to either:

1. Set the maximum chunk size to 512KiB and the maximum block size to 1MiB.
2. Set the maximum block size to 2MiB and the maximum chunk size to 1MiB.

Given that the maximum chunk size is already 1MiB and that's a nice, even, round number, I'm leaning towards keeping 1MiB as the maximum chunk size.

Reply:
> have we made any progress specing maximum block sizes?

yes and no. yes, in that we resolved that IPLD doesn’t have one and that the max size is going to be imposed by transport and storage layers.

the recommended max size is 1meg because, and I think I’m quoting you from memory “nothing bad is going to happen if you keep blocks under 1meg.” 🤓

that said, we’ve already identified some corners where we need the `dag-cbor` decoder to handle much larger blocks and therefor can’t have the codec force the limit. the latest one is the CAR header, which will get quite large if you put a lot of root nodes in it.

Reply:
> Given that the maximum chunk size is already 1MiB

Ah I see. It's basically "both", is why we are talking past each other. For data blocks the 1MiB is indeed counted towards payload, with wrapping taking more:
```
~/devel/go-ipfs$ for b in bafkreibq4fevl27rgurgnxbp7adh42aqiyd6ouflxhj3gzmcxcxzbh6lla bafybeids7spn7r37szbnm2xm3jhg7invbu6tiq7bjt7nqmaibfddwynkdq; do cmd/ipfs/ | wc -c
dist/ ipfs  util/ 
[@ Tue 18:55:52.188]
[$?==0  /    0.070s]

~/devel/go-ipfs$ for b in bafkreibq4fevl27rgurgnxbp7adh42aqiyd6ouflxhj3gzmcxcxzbh6lla bafybeids7spn7r37szbnm2xm3jhg7invbu6tiq7bjt7nqmaibfddwynkdq; do cmd/ipfs/ | wc -c
dist/ ipfs  util/ 
[@ Tue 18:55:52.188]
[$?==0  /    0.070s]

~/devel/go-ipfs$ for b in bafkreibq4fevl27rgurgnxbp7adh42aqiyd6ouflxhj3gzmcxcxzbh6lla bafybeids7spn7r37szbnm2xm3jhg7invbu6tiq7bjt7nqmaibfddwynkdq; do cmd/ipfs/ipfs block get $b  | wc -c ; done
 1048576
 1048590
```

The same variable is used for deciding **link block** sizing / when a block is considered full, which is calculated using on-wire sizing.

So we would need to bump the number in the importer to e,g 1.1, or 1.25 or even 2 ( which I have no trouble doing, I just want to make sure we converged 😝  )

Reply:
That's why I'm saying "2". Really, it comes down to:

* We shouldn't generate chunks larger than 1MiB.
* We should _shoot_ for blocks less than 1MiB.
* All implementations must support blocks up to 2MiB.
* Bitswap will _reject_ **messages** more than 4MiB. That includes both the block plus the message framing.

We don't need to squeeze as much out of this as we can, we just need to set sane defaults and limits that users can remember/understand. That and we _really_ don't want to end up in a situation where we need just a _little_ wiggle room but don't have any.

Reply:
@Stebalien addressed with last push. Current PR is fully backwards compatible. Will introduce the breaking changes in next PR based on this one.